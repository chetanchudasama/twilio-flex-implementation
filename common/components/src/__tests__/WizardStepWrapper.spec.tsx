/* eslint-disable react/jsx-props-no-spreading */
import { MuiThemeProvider, createMuiTheme } from "@material-ui/core";

import React from "react";
import { create } from "react-test-renderer";
import flushPromises from "flush-promises";
import { mount } from "enzyme";
import {
  WizardStepWrapper,
  WizardStepWrapperProps,
} from "../WizardStepWrapper/WizardStepWrapper";

const core: WizardStepWrapperProps = {
  onNext: jest.fn(),
  onPrevious: jest.fn(),
  nextStep: "Next Step",
  previousStep: "Previous Step",
};
let props: WizardStepWrapperProps = { ...core };

const theme = createMuiTheme({
  typography: {
    useNextVariants: true,
  },
});

describe("WizardStepWrapper", () => {
  beforeEach(() => {
    props = { ...core };
  });

  afterEach(() => {
    jest.clearAllMocks();
    jest.clearAllTimers();
  });

  it("renders correctly", () => {
    const wrapper = create(
      <MuiThemeProvider theme={theme}>
        <WizardStepWrapper {...props} />
      </MuiThemeProvider>
    ).toJSON();
    expect(wrapper).toMatchSnapshot();
  });

  it("when previous button is disabled, it should render correctly", () => {
    props.disablePrevious = true;
    const wrapper = create(
      <MuiThemeProvider theme={theme}>
        <WizardStepWrapper {...props} />
      </MuiThemeProvider>
    ).toJSON();
    expect(wrapper).toMatchSnapshot();
  });

  it("when next button is disabled, it should render correctly", () => {
    props.disableNext = true;
    const wrapper = create(
      <MuiThemeProvider theme={theme}>
        <WizardStepWrapper {...props} />
      </MuiThemeProvider>
    ).toJSON();
    expect(wrapper).toMatchSnapshot();
  });

  it("when action is set hidden, it should render correctly", () => {
    props.hideActions = true;
    const wrapper = create(
      <MuiThemeProvider theme={theme}>
        <WizardStepWrapper {...props} />
      </MuiThemeProvider>
    ).toJSON();
    expect(wrapper).toMatchSnapshot();
  });

  it("when disable previous prop passed as true, previous button should be disabled", async () => {
    props.disablePrevious = true;
    const wrapper = mount(
      <MuiThemeProvider theme={theme}>
        <WizardStepWrapper
          {...props}
          classes={{ previousButton: "previous-btn" }}
        />
      </MuiThemeProvider>
    );
    await flushPromises();

    const previousButton = wrapper.find(".previous-btn");
    expect(previousButton.exists()).toBeTruthy();
    expect(previousButton.last().getElement().props.disabled).toBeTruthy();
  });

  it("when disable next props passed as true, next button should be disabled", async () => {
    props.disableNext = true;
    const wrapper = mount(
      <MuiThemeProvider theme={theme}>
        <WizardStepWrapper {...props} classes={{ nextButton: "next-btn" }} />
      </MuiThemeProvider>
    );
    await flushPromises();

    const nextButton = wrapper.find(".next-btn");
    expect(nextButton.exists()).toBeTruthy();
    expect(nextButton.last().getElement().props.disabled).toBeTruthy();
  });

  it("when component is an intermediate step, next and previous buttons should not be disabled", async () => {
    const wrapper = mount(
      <MuiThemeProvider theme={theme}>
        <WizardStepWrapper
          {...props}
          classes={{ nextButton: "next-btn", previousButton: "previous-btn" }}
        />
      </MuiThemeProvider>
    );
    await flushPromises();

    const nextButton = wrapper.find(".next-btn");
    const previousButton = wrapper.find(".previous-btn");

    expect(nextButton.exists()).toBeTruthy();
    expect(previousButton.exists()).toBeTruthy();
    expect(nextButton.last().getElement().props.disabled).toBeFalsy();
    expect(previousButton.last().getElement().props.disabled).toBeFalsy();
  });

  it("when action is set hidden, next and previous buttons should not be visible", async () => {
    props.hideActions = true;
    const wrapper = mount(
      <MuiThemeProvider theme={theme}>
        <WizardStepWrapper
          {...props}
          classes={{ nextButton: "next-btn", previousButton: "previous-btn" }}
        />
      </MuiThemeProvider>
    );
    await flushPromises();

    const nextButton = wrapper.find(".next-btn");
    const previousButton = wrapper.find(".previous-btn");

    expect(nextButton.exists()).toBeFalsy();
    expect(previousButton.exists()).toBeFalsy();
  });

  describe("Next button functionality test", () => {
    it("should call injected onNext event, if not disabled", async () => {
      const onNext = jest.fn();
      props.onNext = onNext;

      const wrapper = mount(
        <MuiThemeProvider theme={theme}>
          <WizardStepWrapper {...props} classes={{ nextButton: "next-btn" }} />
        </MuiThemeProvider>
      );
      await flushPromises();
      const button = wrapper.find(".next-btn").last();
      button.simulate("click");
      await flushPromises();

      expect(onNext).toHaveBeenCalled();
    });

    it("does not call injected onNext method, if disabled", async () => {
      const onNext = jest.fn();
      props.onNext = onNext;
      props.disableNext = true;

      const wrapper = mount(
        <MuiThemeProvider theme={theme}>
          <WizardStepWrapper {...props} classes={{ nextButton: "next-btn" }} />
        </MuiThemeProvider>
      );
      await flushPromises();
      const button = wrapper.find(".next-btn").last();
      button.simulate("click");
      await flushPromises();

      expect(onNext).toHaveBeenCalledTimes(0);
    });
  });

  describe("Previous button functionality test", () => {
    it("should call injected onPrevious event, if not disabled", async () => {
      const onPrevious = jest.fn();
      props.onPrevious = onPrevious;

      const wrapper = mount(
        <MuiThemeProvider theme={theme}>
          <WizardStepWrapper
            {...props}
            classes={{ previousButton: "previous-btn" }}
          />
        </MuiThemeProvider>
      );
      await flushPromises();

      const button = wrapper.find(".previous-btn").last();
      button.simulate("click");
      await flushPromises();

      expect(onPrevious).toHaveBeenCalled();
    });

    it("does not call injected onPrevious method, if disabled", async () => {
      const onPrevious = jest.fn();
      props.onPrevious = onPrevious;
      props.disablePrevious = true;

      const wrapper = mount(
        <MuiThemeProvider theme={theme}>
          <WizardStepWrapper
            {...props}
            classes={{ previousButton: "previous-btn" }}
          />
        </MuiThemeProvider>
      );
      await flushPromises();

      const button = wrapper.find(".previous-btn").last();
      button.simulate("click");
      await flushPromises();

      expect(onPrevious).toHaveBeenCalledTimes(0);
    });
  });
});
