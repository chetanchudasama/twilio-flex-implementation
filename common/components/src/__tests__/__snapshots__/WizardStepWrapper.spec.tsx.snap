// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WizardStepWrapper renders correctly 1`] = `
<div
  className="MuiGrid-item-9 MuiGrid-grid-xs-12-48 WizardStepWrapperContent-root-1"
>
  <div
    className="MuiGrid-container-8 MuiGrid-spacing-xs-16-31 MuiGrid-align-items-xs-center-16 MuiGrid-justify-xs-center-25"
  >
    <div
      className="MuiGrid-item-9 MuiGrid-grid-xs-12-48 WizardStepWrapperContent-content-2"
      style={
        Object {
          "height": "fit-content",
        }
      }
    />
    <div
      className="MuiGrid-item-9 MuiGrid-grid-xs-12-48 actionContainer WizardStepWrapperContent-actionContainer-7"
    >
      <div
        className="MuiGrid-container-8 MuiGrid-spacing-xs-16-31 MuiGrid-align-items-xs-center-16 MuiGrid-justify-xs-space-between-27"
      >
        <div
          className="MuiGrid-item-9"
        >
          <button
            className="MuiButtonBase-root-131 MuiButton-root-105 MuiButton-text-107 MuiButton-textSecondary-109 MuiButton-flat-110 MuiButton-flatSecondary-112 WizardStepWrapperContent-previousButton-4"
            disabled={false}
            onBlur={[Function]}
            onClick={[MockFunction]}
            onContextMenu={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex="0"
            type="button"
          >
            <span
              className="MuiButton-label-106"
            >
              <svg
                aria-hidden="true"
                className="MuiSvgIcon-root-134 WizardStepWrapperContent-arrowBackBtn-5"
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <path
                  d="M11.67 3.87L9.9 2.1 0 12l9.9 9.9 1.77-1.77L3.54 12z"
                />
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
              </svg>
              Previous: Previous Step
            </span>
            <span
              className="MuiTouchRipple-root-143"
            />
          </button>
        </div>
        <div
          className="MuiGrid-item-9"
        >
          <button
            className="MuiButtonBase-root-131 MuiButton-root-105 MuiButton-contained-116 MuiButton-containedSecondary-118 MuiButton-raised-119 MuiButton-raisedSecondary-121 WizardStepWrapperContent-nextButton-3"
            disabled={false}
            onBlur={[Function]}
            onClick={[MockFunction]}
            onContextMenu={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex="0"
            type="button"
          >
            <span
              className="MuiButton-label-106"
            >
              Next: Next Step
              <svg
                aria-hidden="true"
                className="MuiSvgIcon-root-134 WizardStepWrapperContent-arrowForwardBtn-6"
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <path
                  d="M5.88 4.12L13.76 12l-7.88 7.88L8 22l10-10L8 2z"
                />
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
              </svg>
            </span>
            <span
              className="MuiTouchRipple-root-143"
            />
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`WizardStepWrapper when action is set hidden, it should render correctly 1`] = `
<div
  className="MuiGrid-item-9 MuiGrid-grid-xs-12-48 WizardStepWrapperContent-root-1"
>
  <div
    className="MuiGrid-container-8 MuiGrid-spacing-xs-16-31 MuiGrid-align-items-xs-center-16 MuiGrid-justify-xs-center-25"
  >
    <div
      className="MuiGrid-item-9 MuiGrid-grid-xs-12-48 WizardStepWrapperContent-content-2"
      style={
        Object {
          "height": "fit-content",
        }
      }
    />
  </div>
</div>
`;

exports[`WizardStepWrapper when next button is disabled, it should render correctly 1`] = `
<div
  className="MuiGrid-item-9 MuiGrid-grid-xs-12-48 WizardStepWrapperContent-root-1"
>
  <div
    className="MuiGrid-container-8 MuiGrid-spacing-xs-16-31 MuiGrid-align-items-xs-center-16 MuiGrid-justify-xs-center-25"
  >
    <div
      className="MuiGrid-item-9 MuiGrid-grid-xs-12-48 WizardStepWrapperContent-content-2"
      style={
        Object {
          "height": "fit-content",
        }
      }
    />
    <div
      className="MuiGrid-item-9 MuiGrid-grid-xs-12-48 actionContainer WizardStepWrapperContent-actionContainer-7"
    >
      <div
        className="MuiGrid-container-8 MuiGrid-spacing-xs-16-31 MuiGrid-align-items-xs-center-16 MuiGrid-justify-xs-space-between-27"
      >
        <div
          className="MuiGrid-item-9"
        >
          <button
            className="MuiButtonBase-root-131 MuiButton-root-105 MuiButton-text-107 MuiButton-textSecondary-109 MuiButton-flat-110 MuiButton-flatSecondary-112 WizardStepWrapperContent-previousButton-4"
            disabled={false}
            onBlur={[Function]}
            onClick={[MockFunction]}
            onContextMenu={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex="0"
            type="button"
          >
            <span
              className="MuiButton-label-106"
            >
              <svg
                aria-hidden="true"
                className="MuiSvgIcon-root-134 WizardStepWrapperContent-arrowBackBtn-5"
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <path
                  d="M11.67 3.87L9.9 2.1 0 12l9.9 9.9 1.77-1.77L3.54 12z"
                />
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
              </svg>
              Previous: Previous Step
            </span>
            <span
              className="MuiTouchRipple-root-143"
            />
          </button>
        </div>
        <div
          className="MuiGrid-item-9"
        >
          <button
            className="MuiButtonBase-root-131 MuiButtonBase-disabled-132 MuiButton-root-105 MuiButton-contained-116 MuiButton-containedSecondary-118 MuiButton-raised-119 MuiButton-raisedSecondary-121 MuiButton-disabled-125 WizardStepWrapperContent-nextButton-3"
            disabled={true}
            onBlur={[Function]}
            onClick={[MockFunction]}
            onContextMenu={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex="-1"
            type="button"
          >
            <span
              className="MuiButton-label-106"
            >
              Next
              <svg
                aria-hidden="true"
                className="MuiSvgIcon-root-134 WizardStepWrapperContent-arrowForwardBtn-6"
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <path
                  d="M5.88 4.12L13.76 12l-7.88 7.88L8 22l10-10L8 2z"
                />
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
              </svg>
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`WizardStepWrapper when previous button is disabled, it should render correctly 1`] = `
<div
  className="MuiGrid-item-9 MuiGrid-grid-xs-12-48 WizardStepWrapperContent-root-1"
>
  <div
    className="MuiGrid-container-8 MuiGrid-spacing-xs-16-31 MuiGrid-align-items-xs-center-16 MuiGrid-justify-xs-center-25"
  >
    <div
      className="MuiGrid-item-9 MuiGrid-grid-xs-12-48 WizardStepWrapperContent-content-2"
      style={
        Object {
          "height": "fit-content",
        }
      }
    />
    <div
      className="MuiGrid-item-9 MuiGrid-grid-xs-12-48 actionContainer WizardStepWrapperContent-actionContainer-7"
    >
      <div
        className="MuiGrid-container-8 MuiGrid-spacing-xs-16-31 MuiGrid-align-items-xs-center-16 MuiGrid-justify-xs-space-between-27"
      >
        <div
          className="MuiGrid-item-9"
        >
          <button
            className="MuiButtonBase-root-131 MuiButtonBase-disabled-132 MuiButton-root-105 MuiButton-text-107 MuiButton-textSecondary-109 MuiButton-flat-110 MuiButton-flatSecondary-112 MuiButton-disabled-125 WizardStepWrapperContent-previousButton-4"
            disabled={true}
            onBlur={[Function]}
            onClick={[MockFunction]}
            onContextMenu={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex="-1"
            type="button"
          >
            <span
              className="MuiButton-label-106"
            >
              <svg
                aria-hidden="true"
                className="MuiSvgIcon-root-134 WizardStepWrapperContent-arrowBackBtn-5"
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <path
                  d="M11.67 3.87L9.9 2.1 0 12l9.9 9.9 1.77-1.77L3.54 12z"
                />
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
              </svg>
              Previous
            </span>
          </button>
        </div>
        <div
          className="MuiGrid-item-9"
        >
          <button
            className="MuiButtonBase-root-131 MuiButton-root-105 MuiButton-contained-116 MuiButton-containedSecondary-118 MuiButton-raised-119 MuiButton-raisedSecondary-121 WizardStepWrapperContent-nextButton-3"
            disabled={false}
            onBlur={[Function]}
            onClick={[MockFunction]}
            onContextMenu={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex="0"
            type="button"
          >
            <span
              className="MuiButton-label-106"
            >
              Next: Next Step
              <svg
                aria-hidden="true"
                className="MuiSvgIcon-root-134 WizardStepWrapperContent-arrowForwardBtn-6"
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <path
                  d="M5.88 4.12L13.76 12l-7.88 7.88L8 22l10-10L8 2z"
                />
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
              </svg>
            </span>
            <span
              className="MuiTouchRipple-root-143"
            />
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
